custom-prompt/custom-prompt.cc

```diff
@@ -163,6 +171,288 @@ void Interval::print_long(std::ostream& ostream) const
  *****************************************************************************/
 extern "C" long long unsigned get_active_wid(void);
 
+/******************************************************************************
+ * Store information about the current Git repository.
+ *****************************************************************************/
+class GitRepository
+{
+private:
+    C::git_repository* repo;
+    bool bare, detached;
+    std::filesystem::path gitdir;
+    C::git_reference* ref;
+    C::git_oid const* oid;
+    std::string description, tag;
+    std::string state;
+    bool dirty, staged, untracked;
+
+public:
+    GitRepository(void);
+    std::string get_information(void);
+
+private:
+    void establish_description(void);
+    void establish_tag(void);
+    void establish_state(void);
+    void establish_dirty_staged_untracked(void);
+    // These are static methods because otherwise, their signatures do not
+    // match the required signatures for use as callback functions.
+    static int update_tag(char const*, C::git_oid*, void*);
+    static int update_dirty_staged_untracked(char const*, unsigned, void*);
+};
+
+/******************************************************************************
+ * Read the current Git repository.
+ *****************************************************************************/
+GitRepository::GitRepository(void) :
+    repo(nullptr), bare(false), detached(false), ref(nullptr), oid(nullptr), dirty(false), staged(false),
+    untracked(false)
+{
+    if (C::git_libgit2_init() <= 0)
+    {
+        return;
+    }
+    if (C::git_repository_open_ext(&this->repo, ".", 0, nullptr) != 0)
+    {
+        return;
+    }
+    this->bare = C::git_repository_is_bare(this->repo);
+    this->detached = C::git_repository_head_detached(this->repo);
+    this->gitdir = C::git_repository_path(this->repo);
+    this->establish_description();
+    this->establish_tag();
+    this->establish_state();
+    this->establish_dirty_staged_untracked();
+}
+
+/******************************************************************************
+ * Obtain a human-readable description of the working tree of the current Git
+ * repository. This shall be the name of the current branch if it is available.
+ * Otherwise, it shall be the object ID of the most recent commit.
+ *****************************************************************************/
+void GitRepository::establish_description(void)
+{
+    if (C::git_repository_head(&this->ref, this->repo) == 0)
+    {
+        // According to the documentation, this retrieves the reference object
+        // ID only if the reference is direct. However, I observed that it does
+        // so even if the reference is symbolic (i.e. if we are on a branch).
+        // There is no harm in leaving it here because if it fails, it will
+        // just return a null pointer.
+        this->oid = C::git_reference_target(this->ref);
```

Comment:
#### Suggestion

`git_reference_name_to_id` instead? No need to store `ref` then.

Reply:
No. Need the reference to determine the branch name.